homeassistant/components/tuya/__init__.py

```diff
@@ -107,86 +123,144 @@ def load_devices(device_list):
         # Add new discover device.
-        device_list = tuya.get_all_devices()
-        load_devices(device_list)
+        device_list = await hass.async_add_executor_job(_get_updated_devices)
+        await async_load_devices(device_list)
         # Delete not exist device.
         newlist_ids = []
         for device in device_list:
             newlist_ids.append(device.object_id())
         for dev_id in list(hass.data[DOMAIN]["entities"]):
             if dev_id not in newlist_ids:
-                dispatcher_send(hass, SIGNAL_DELETE_ENTITY, dev_id)
+                async_dispatcher_send(hass, SIGNAL_DELETE_ENTITY, dev_id)
                 hass.data[DOMAIN]["entities"].pop(dev_id)
 
-    track_time_interval(hass, poll_devices_update, timedelta(minutes=5))
+    hass.data[DOMAIN][TUYA_TRACKER] = async_track_time_interval(
+        hass, async_poll_devices_update, timedelta(minutes=5)
+    )
 
-    hass.services.register(DOMAIN, SERVICE_PULL_DEVICES, poll_devices_update)
+    hass.services.async_register(
+        DOMAIN, SERVICE_PULL_DEVICES, async_poll_devices_update
+    )
 
-    def force_update(call):
+    async def async_force_update(call):
         """Force all devices to pull data."""
-        dispatcher_send(hass, SIGNAL_UPDATE_ENTITY)
+        async_dispatcher_send(hass, SIGNAL_UPDATE_ENTITY)
 
-    hass.services.register(DOMAIN, SERVICE_FORCE_UPDATE, force_update)
+    hass.services.async_register(DOMAIN, SERVICE_FORCE_UPDATE, async_force_update)
 
     return True
 
 
+async def async_unload_entry(hass, entry):
+    """Unloading the Tuya platforms."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(
+                    entry, component.split(".", 1)[0]
+                )
+                for component in hass.data[DOMAIN][ENTRY_IS_SETUP]
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN][ENTRY_IS_SETUP] = set()
+        hass.data[DOMAIN][TUYA_TRACKER]()
+        hass.data[DOMAIN][TUYA_TRACKER] = None
+        hass.data[DOMAIN][TUYA_DATA] = None
+        hass.services.async_remove(DOMAIN, SERVICE_FORCE_UPDATE)
+        hass.services.async_remove(DOMAIN, SERVICE_PULL_DEVICES)
+        hass.data.pop(DOMAIN)
+
+    return unload_ok
+
+
 class TuyaDevice(Entity):
     """Tuya base device."""
 
-    def __init__(self, tuya):
+    def __init__(self, tuya, platform):
         """Init Tuya devices."""
-        self.tuya = tuya
+        self._tuya = tuya
+        self._platform = platform
 
     async def async_added_to_hass(self):
         """Call when entity is added to hass."""
-        dev_id = self.tuya.object_id()
+        dev_id = self._tuya.object_id()
         self.hass.data[DOMAIN]["entities"][dev_id] = self.entity_id
         async_dispatcher_connect(self.hass, SIGNAL_DELETE_ENTITY, self._delete_callback)
         async_dispatcher_connect(self.hass, SIGNAL_UPDATE_ENTITY, self._update_callback)
```

Comment:
```suggestion
        async_dispatcher_connect(self.hass, SIGNAL_UPDATE_ENTITY, self.async_write_ha_state)
```

And than delete `self._update_callback`

Reply:
As I know `async_write_ha_state` do not force state update

Reply:
Right, I missed this wasn't coming from an update but was requesting one.